LegendaryExplorer/LegendaryExplorer/Tools/Sequence Editor/Experiments/SequenceEditorExperimentsE.cs

```diff
@@ -0,0 +1,247 @@
+﻿
+using LegendaryExplorer.Dialogs;
+using LegendaryExplorerCore.Helpers;
+using LegendaryExplorerCore.Kismet;
+using LegendaryExplorerCore.Matinee;
+using LegendaryExplorerCore.Packages;
+using LegendaryExplorerCore.Unreal;
+using LegendaryExplorerCore.Unreal.ObjectInfo;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+using static LegendaryExplorerCore.Kismet.SeqTools;
+
+namespace LegendaryExplorer.Tools.Sequence_Editor.Experiments {
+    /// <summary>
+    /// Experiments in Sequence Editor (Exkywor's stuff)
+    /// </summary>
+    public static class SequenceEditorExperimentsE {
+        /// <summary>
+        /// Updates the variable links of the selected Interp. Removes links that no longer have a matching InterpGroup,
+        /// adds links for new groups, and keeps groups that remaing the same.
+        /// </summary>
+        /// <param name="sew">Sequence Editor window.</param>
+        /// <param name="inLoop">Whether we're updating in a loop. Used to avoid double-checking what the caller has already checked.</param>
+        /// <param name="interp">Interp to update. Used to pass an interp gotten from a loop, rather than being a selected object.</param>
+        public static void UpdateVarLinks(SequenceEditorWPF sew, bool inLoop = false, ExportEntry interp = null) {
+            if (interp == null) {
+                if (sew.Pcc == null || sew.SelectedItem == null || sew.SelectedItem.Entry == null) { return; }
+
+                if (!sew.SelectedObjects.HasExactly(1)) {
+                    ShowError("Select only one Interp object");
+                    return;
+                }
+
+                interp = sew.SelectedObjects[0].Export;
+                if (interp.ClassName != "SeqAct_Interp") {
+                    ShowError("Selected object is not a SeqAct_Interp");
+                    return;
+                }
+            }
+
+            // We get a list of StructProperties instead of VarLinkInfo because we want to keep existing ones intact
+            List<StructProperty> varLinks = interp.GetProperty<ArrayProperty<StructProperty>>("VariableLinks").ToList();
+            if (!inLoop) {
+                if (varLinks == null) {
+                    ShowError("The selected Interp contains no VariableLinks");
+                    return;
+                }
+            }
+
+            List<StructProperty> dataLinks = varLinks.Where(link => link.GetProp<StrProperty>("LinkDesc").Value == "Data").ToList();
+            if (!inLoop) {
+                if (!dataLinks.Any()) {
+                    ShowError("The selected Interp contains no Data variable link");
+                    return;
+                }
+            }
+
+            ObjectProperty dataObj = dataLinks.First().GetProp<ArrayProperty<ObjectProperty>>("LinkedVariables").FirstOrDefault();
```

Comment:
The call to `First()` will throw an exception if dataLinks is empty. You early-return a couple lines earlier if it's empty, but only when `inLoop` is false.

Now, currently, that's technically not a problem, since if inLoop is false the checks will have been done in the calling function. However, it would be all too easy in the future to make a change to the checks in one function but forget to do it in the other. Better to just do the checks twice, as none of them are computationally significant.

Reply:
Aaah, alright. I'll add that. I was originally going to do put them twice as suggested, but thought it'd be bad to copy paste the same code too much